lib/brainstem/api_docs/presenter.rb

```diff
@@ -47,88 +45,91 @@ def initialize(atlas, options = {})
         yield self if block_given?
       end
 
-
       def suggested_filename(format)
         filename_pattern
           .gsub('{{name}}', target_class.to_s.underscore)
           .gsub('{{extension}}', extension)
       end
 
-
       def suggested_filename_link(format)
         filename_link_pattern
           .gsub('{{name}}', target_class.to_s.underscore)
           .gsub('{{extension}}', extension)
       end
 
-
       attr_accessor :atlas
 
-
       def extension
         @extension ||= Brainstem::ApiDocs.output_extension
       end
 
-
       def filename_pattern
         @filename_pattern ||= Brainstem::ApiDocs.presenter_filename_pattern
       end
 
-
       def filename_link_pattern
         @filename_link_pattern ||= Brainstem::ApiDocs.presenter_filename_link_pattern
       end
 
-
       delegate :configuration => :const
       delegate :find_by_class => :atlas
 
-
       def nodoc?
         configuration[:nodoc]
       end
 
-
       def title
         contextual_documentation(:title) || const.to_s.demodulize
       end
 
-
       def brainstem_keys
         const.possible_brainstem_keys.to_a.sort
       end
 
-
       def description
         contextual_documentation(:description) || ""
       end
 
-
       def valid_fields(fields = configuration[:fields])
         fields.to_h.reject do |_, field|
           invalid_field?(field) || (nested_field?(field) && valid_fields_in(field).none?)
         end
       end
       alias_method :valid_fields_in, :valid_fields
 
-
       def invalid_field?(field)
         field.options[:nodoc]
       end
 
-
       def nested_field?(field)
         field.respond_to?(:configuration)
       end
 
+      def optional_field?(field)
+        field.options[:optional]
+      end
+
+      def optional_field_names(fields = configuration[:fields], buffer = [])
+        fields.to_h.select do |field_name, field_config|
```

Comment:
Looks like this can just be an `each`

Reply:
`select` is correct here. I am ignoring fields that are `nodoc`

Reply:
Shouldn't `select` return a boolean to filter down the hash? 

I think this only works because we're mutating the `buffer` variable instead using the return value of `select`. 

https://ruby-doc.org/core-2.2.0/Array.html#method-i-select

Reply:
Fixed. Don't know what I was thinking while implementing / replying to your comment :)